plugins/inputs/t128_tank/reader.go

```diff
@@ -0,0 +1,354 @@
+	}
+	rawLines = bytes.TrimSpace(rawLines)
+	lines := bytes.Split(rawLines, []byte("\n"))
+	for _, rawLine := range lines {
+		line := string(rawLine)
+
+		if len(line) == 0 {
+			continue
+		}
+		message, err := ParseTankLine(rawLine)
+		if err != nil {
+			fmt.Printf("parse failure for topic %s: %v", topic, err)
+			continue
+		}
+
+		if message == nil {
+			if isBoundaryFault, nextAvailabelIndex := isBoundaryFault(line); isBoundaryFault {
+				return messages, fmt.Errorf(
+					"parsed boundary fault: %w",
+					&boundaryFault{nextAvailableIndex: nextAvailabelIndex},
+				)
+			}
+
+			fmt.Printf("skipping %s line because regex matching failed: %s", topic, line)
+			continue
+		}
+
+		message.Message = bytes.TrimRight(message.Message, "\n")
+		messages = append(messages, message)
+	}
+
+	return messages, err
+}
+
+func ParseTankLine(line []byte) (*IndexedMessage, error) {
+	matches := indexRegex.FindSubmatch(line)
+	if len(matches) != 3 {
+		return nil, errors.New("invalid line format")
+	}
+
+	index, err := newIndex(string(matches[1]))
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse index %s, skipping: %s", matches[1], err)
+	}
+
+	message := matches[2]
+	return &IndexedMessage{
+		Message: message,
+		Index:   index,
+	}, nil
+}
+
+func isBoundaryFault(line string) (bool, index) {
+	matches := boundaryFaultRegex.FindStringSubmatch(line)
+	if len(matches) != 2 {
+		return false, StartIndex
+	}
+
+	nextIndex, err := newIndex(matches[1])
+	if err != nil {
+		fmt.Printf("unable to parse index after detecting boundary fault: %s", err)
+		return false, StartIndex
+	}
+
+	return true, nextIndex
+}
+
+func (r *Reader) readFromTank(readCtx context.Context, startingIndex index) (err error) {
+	cmdArgs := []string{
+		"/opt/128technology/bin/tank-cli",
+		"-b",
+		fmt.Sprintf("%s:%d", r.tankAddress, r.tankPort),
+		"-t",
+		r.topic,
+		"consume",
+		"-F",
+		"seqnum,content",
+		startingIndex.string(),
+	}
+
+	cmd := r.tankReadCmdCtx(readCtx, cmdArgs[0], cmdArgs[1:]...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("unable to pipe %s tank output: %s", r.topic, err)
+	}
+
+	tankReader := bufio.NewReader(stdout)
+
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("unable to start %s tank stream: %w", r.topic, err)
+	}
+
+	var parseErr error
+	var messages []*IndexedMessage
+
+	for {
+		select {
+		case <-readCtx.Done():
+			fmt.Printf("%s read routine done", r.topic)
+			return nil
+		default:
+		}
+
+		messages, parseErr = parseLines(tankReader, r.topic)
+		if parseErr != nil {
+			fmt.Printf("encountered error while parsing lines in %s output, will not attempt to parse more lines: %v", r.topic, parseErr)
+			return parseErr
+		}
+		fmt.Println("After Parse Line inside read from tank")
+		for _, message := range messages {
+			if message.IsValid() {
+				select {
+				case <-readCtx.Done():
+					return nil
+				case r.sendChan <- *message:
```

Comment:
Change this channel to take all the messages at once. No need to send them individually.